File: 695.max-area-of-island.py
# @leet start
from collections import deque 

class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        m, n = len(grid), len(grid[0])
        max_size = 0

        for y in range(m):
            for x in range(n):
                if grid[y][x] == 1:
                    curr_size = 0
                    todo = deque()
                    todo.append((x,y))

                    while len(todo) > 0:
                        xx, yy = todo.popleft()

                        if grid[yy][xx] == 0:
                            continue
                        
                        grid[yy][xx] = 0
                        curr_size += 1

                        if xx - 1 >= 0:
                            todo.append((xx - 1, yy))

                        if xx + 1 < n:
                            todo.append((xx + 1, yy))

                        if yy - 1 >= 0:
                            todo.append((xx, yy - 1))

                        if yy + 1 < m:
                            todo.append((xx, yy + 1))

                    max_size = max(max_size, curr_size)
        return max_size
        
# @leet end

File: 695.max-area-of-island.ts
// @leet start
function maxAreaOfIsland(grid: number[][]): number {
	let result = 0;
	let m = grid.length;
	let n = grid[0].length;

	const dfs = (x: number, y: number, size: number): number => {
		if (grid[y][x] == 0) {
			return size;
		}

		grid[y][x] = 0;
		size++;

		if (y + 1 < m) {
			size = dfs(x, y + 1, size);
		}

		if (y - 1 >= 0) {
			size = dfs(x, y - 1, size);
		}

		if (x + 1 < n) {
			size = dfs(x + 1, y, size);
		}

		if (x - 1 >= 0) {
			size = dfs(x - 1, y, size);
		}

		return size;
	}

	for (let y = 0; y < m; y++) {
		for (let x = 0; x < n; x++) {
			if (grid[y][x] == 1) {
				result = Math.max(result, dfs(x, y, 0));
			}
		}
	}

	return result;
    
};
// @leet end
